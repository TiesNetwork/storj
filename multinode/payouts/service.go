// Copyright (C) 2021 Storj Labs, Inc.
// See LICENSE for copying information.

package payouts

import (
	"context"

	"github.com/spacemonkeygo/monkit/v3"
	"github.com/zeebo/errs"
	"go.uber.org/zap"

	"storj.io/common/rpc"
	"storj.io/common/storj"
	"storj.io/storj/multinode/nodes"
	"storj.io/storj/private/multinodepb"
)

var (
	mon = monkit.Package()
	// Error is an error class for payouts service error.
	Error = errs.Class("payouts service error")
)

// Service exposes all payouts related logic.
//
// architecture: Service
type Service struct {
	log    *zap.Logger
	dialer rpc.Dialer
	nodes  nodes.DB
}

// NewService creates new instance of Service.
func NewService(log *zap.Logger, dialer rpc.Dialer, nodes nodes.DB) *Service {
	return &Service{
		log:    log,
		dialer: dialer,
		nodes:  nodes,
	}
}

// GetAllNodesAllTimeEarned retrieves all nodes earned amount for all time.
func (service *Service) GetAllNodesAllTimeEarned(ctx context.Context) (earned int64, err error) {
	defer mon.Task()(&ctx)(&err)

	storageNodes, err := service.nodes.List(ctx)
	if err != nil {
		return 0, Error.Wrap(err)
	}

	for _, node := range storageNodes {
		amount, err := service.getAmount(ctx, node)
		if err != nil {
			service.log.Error("failed to getAmount", zap.Error(err))
			continue
		}

		earned += amount
	}

	return earned, nil
}

// GetAllNodesEarnedOnSatellite retrieves all nodes earned amount for all time per satellite.
func (service *Service) GetAllNodesEarnedOnSatellite(ctx context.Context) (earned []SatelliteSummary, err error) {
	defer mon.Task()(&ctx)(&err)

	storageNodes, err := service.nodes.List(ctx)
	if err != nil {
		return nil, Error.Wrap(err)
	}

	var listSatellites storj.NodeIDList
	var listNodesEarnedPerSatellite []multinodepb.EarnedPerSatelliteResponse

	for _, node := range storageNodes {
		earnedPerSatellite, err := service.getEarnedOnSatellite(ctx, node)
		if err != nil {
			service.log.Error("failed to getEarnedFromSatellite", zap.Error(err))
			continue
		}

		listNodesEarnedPerSatellite = append(listNodesEarnedPerSatellite, earnedPerSatellite)
		for i := 0; i < len(earnedPerSatellite.EarnedSatellite); i++ {
			listSatellites = append(listSatellites, earnedPerSatellite.EarnedSatellite[i].SatelliteId)
		}
	}

	if listSatellites == nil {
		return []SatelliteSummary{}, nil
	}

	uniqueSatelliteIDs := listSatellites.Unique()
	for t := 0; t < len(uniqueSatelliteIDs); t++ {
		earned = append(earned, SatelliteSummary{
			SatelliteID: uniqueSatelliteIDs[t],
		})
	}

	for i := 0; i < len(listNodesEarnedPerSatellite); i++ {
		singleNodeEarnedPerSatellite := listNodesEarnedPerSatellite[i].EarnedSatellite
		for j := 0; j < len(singleNodeEarnedPerSatellite); j++ {
			for k := 0; k < len(earned); k++ {
				if singleNodeEarnedPerSatellite[j].SatelliteId == earned[k].SatelliteID {
					earned[k].Earned += singleNodeEarnedPerSatellite[j].Total
				}
			}
		}
	}

	return earned, nil
}

func (service *Service) getAmount(ctx context.Context, node nodes.Node) (_ int64, err error) {
	conn, err := service.dialer.DialNodeURL(ctx, storj.NodeURL{
		ID:      node.ID,
		Address: node.PublicAddress,
	})
	if err != nil {
		return 0, Error.Wrap(err)
	}

	defer func() {
		err = errs.Combine(err, conn.Close())
	}()

	payoutClient := multinodepb.NewDRPCPayoutClient(conn)
	header := &multinodepb.RequestHeader{
		ApiKey: node.APISecret,
	}

	amount, err := payoutClient.Earned(ctx, &multinodepb.EarnedRequest{Header: header})
	if err != nil {
		return 0, Error.Wrap(err)
	}

	return amount.Total, nil
}

func (service *Service) getEarnedOnSatellite(ctx context.Context, node nodes.Node) (_ multinodepb.EarnedPerSatelliteResponse, err error) {
	conn, err := service.dialer.DialNodeURL(ctx, storj.NodeURL{
		ID:      node.ID,
		Address: node.PublicAddress,
	})
	if err != nil {
		return multinodepb.EarnedPerSatelliteResponse{}, Error.Wrap(err)
	}

	defer func() {
		err = errs.Combine(err, conn.Close())
	}()

	payoutClient := multinodepb.NewDRPCPayoutClient(conn)
	header := &multinodepb.RequestHeader{
		ApiKey: node.APISecret,
	}

	response, err := payoutClient.EarnedPerSatellite(ctx, &multinodepb.EarnedPerSatelliteRequest{Header: header})
	if err != nil {
		return multinodepb.EarnedPerSatelliteResponse{}, Error.Wrap(err)
	}

	return *response, nil
}
