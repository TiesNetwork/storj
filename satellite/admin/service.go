package admin

import (
	"storj.io/storj/satellite/accounting"
	"storj.io/storj/satellite/admin/service"
	"storj.io/storj/satellite/console"
)

// ServiceConfig is a configuration for administration service
type ServiceConfig service.Config

// NewService returns new instance of Service.
func NewService(consoleDB console.DB, projectDB accounting.ProjectAccounting, storagenodeDB accounting.StoragenodeAccounting, adminDB DB, projectUsage *accounting.Service, config *ServiceConfig) *service.Service {
	return service.NewService(consoleDB, projectDB, storagenodeDB, adminDB.Nodes(), projectUsage, (*service.Config)(config))
}
