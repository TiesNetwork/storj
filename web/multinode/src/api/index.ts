// Copyright (C) 2021 Storj Labs, Inc.
// See LICENSE for copying information.

/**
 * ErrorUnauthorized is a custom error type for performing unauthorized operations.
 */
export class UnauthorizedError extends Error {
    public constructor(message: string = 'authorization required') {
        super(message);
    }
}

/**
 * BadRequestError is a custom error type for performing bad request.
 */
export class BadRequestError extends Error {
    public constructor(message: string = 'bad request') {
        super(message);
    }
}

/**
 * InternalError is a custom error type for internal server error.
 */
export class InternalError extends Error {
    public constructor(message: string = 'internal server error') {
        super(message);
    }
}
