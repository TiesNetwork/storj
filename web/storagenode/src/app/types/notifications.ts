// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { NotificationIcon } from '@/app/utils/notificationIcons';
import { Notification, NotificationTypes } from '@/storagenode/notifications/notifications';

/**
 * Holds all notifications module state.
 */
export class NotificationsState {
    public latestNotifications: UINotification[] = [];

    public constructor(
        public notifications: UINotification[] = [],
        public pageCount: number = 0,
        public unreadCount: number = 0,
    ) { }
}

/**
 * Describes notification entity.
 */
export class UINotification {
    public icon: NotificationIcon;
    public isRead: boolean;
    public id: string;
    public senderId: string;
    public type: NotificationTypes;
    public title: string;
    public message: string;
    public readAt: Date | null;
    public createdAt: Date = new Date();

    public constructor(notification: Partial<UINotification> = new Notification()) {
        Object.assign(this, notification);
        this.setIcon();
        this.isRead = !!this.readAt;
    }

    /**
     * dateLabels formats createdAt into more informative strings.
     */
    public get dateLabel(): string {
        const differenceInSeconds = (Math.trunc(new Date().getTime()) - Math.trunc(new Date(this.createdAt).getTime())) / 1000;

        switch (true) {
            case differenceInSeconds < 60:
                return 'Just now';
            case differenceInSeconds < 3600:
                return `${(differenceInSeconds / 60).toFixed(0)} minute(s) ago`;
            case differenceInSeconds < 86400:
                return `${(differenceInSeconds / 3600).toFixed(0)} hour(s) ago`;
            case differenceInSeconds < 86400 * 2:
                return `Yesterday`;
            default:
                return new Date(this.createdAt).toDateString();
        }
    }

    /**
     * markAsRead mark notification as read on UI.
     */
    public markAsRead(): void {
        this.isRead = true;
    }

    /**
     * setIcon selects notification icon depends on type.
     */
    private setIcon(): void {
        switch (this.type) {
            case NotificationTypes.AuditCheckFailure:
                this.icon = NotificationIcon.FAIL;
                break;
            case NotificationTypes.Disqualification:
                this.icon = NotificationIcon.SOFTWARE_UPDATE;
                break;
            case NotificationTypes.Suspension:
                this.icon = NotificationIcon.SUSPENDED;
                break;
            default:
                this.icon = NotificationIcon.INFO;
        }
    }
}
