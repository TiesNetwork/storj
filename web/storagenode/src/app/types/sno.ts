// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import {
    BandwidthUsed,
    Checks,
    EgressUsed,
    IngressUsed,
    Node,
    SatelliteInfo,
    SatelliteScores,
    Stamp,
    Utilization,
} from '@/storagenode/sno/sno';

/**
 * Holds all node module state.
 */
export class StorageNodeState {
    public info: Node = new Node();
    public utilization: Utilization = new Utilization();
    public satellites: SatelliteInfo[] = [];
    public disqualifiedSatellites: SatelliteInfo[] = [];
    public suspendedSatellites: SatelliteInfo[] = [];
    public selectedSatellite: SatelliteInfo = new SatelliteInfo();
    public bandwidthChartData: BandwidthUsed[] = [];
    public egressChartData: EgressUsed[] = [];
    public ingressChartData: IngressUsed[] = [];
    public storageChartData: Stamp[] = [];
    public storageSummary: number = 0;
    public bandwidthSummary: number = 0;
    public egressSummary: number = 0;
    public ingressSummary: number = 0;
    public satellitesScores: SatelliteScores[] = [];
    public audits: SatelliteScores = new SatelliteScores();
}
