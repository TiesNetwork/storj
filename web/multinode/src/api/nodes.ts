// Copyright (C) 2021 Storj Labs, Inc.
// See LICENSE for copying information.

import { BadRequestError, InternalError, UnauthorizedError } from '@/api/index';
import { CreateNodeFields, Node, NodeURL } from '@/nodes';
import { HttpClient } from '@/private/http/client';

/**
 * client for nodes controller of MND api.
 */
export class NodesClient {
    private readonly http: HttpClient = new HttpClient();
    private readonly ROOT_PATH: string = '/api/v0/nodes';

    /**
     * handles node addition.
     *
     * @param node - node to add.
     *
     * @throws {@link BadRequestError}
     * This exception is thrown if the input is not a valid.
     *
     * @throws {@link UnauthorizedError}
     * Thrown if the auth cookie is missing or invalid.
     *
     * @throws {@link InternalError}
     * Thrown if something goes wrong on server side.
     */
    public async add(node: CreateNodeFields): Promise<void> {
        const path = `${this.ROOT_PATH}`;
        const response = await this.http.post(path, JSON.stringify(node));

        if (!response.ok) {
            await this.handleError(response);
        }
    }

    /**
     * returns list of node infos.
     *
     * @throws {@link BadRequestError}
     * This exception is thrown if the input is not a valid.
     *
     * @throws {@link UnauthorizedError}
     * Thrown if the auth cookie is missing or invalid.
     *
     * @throws {@link InternalError}
     * Thrown if something goes wrong on server side.
     */
    public async list(): Promise<Node[]> {
        const path = `${this.ROOT_PATH}/infos`;
        const response = await this.http.get(path);

        if (!response.ok) {
            await this.handleError(response);
        }

        const nodeListJson = await response.json();

        return nodeListJson.map(node => new Node(
            node.id,
            node.name,
            node.version,
            new Date(node.lastContact),
            node.diskSpaceUsed,
            node.diskSpaceLeft,
            node.bandwidthUsed,
            0,
            0,
            0,
            node.totalEarned,
        ));
    }

    /**
     * returns list of node infos by satellite.
     *
     * @param satelliteId - id of the satellite.
     *
     * @throws {@link BadRequestError}
     * This exception is thrown if the input is not a valid.
     *
     * @throws {@link UnauthorizedError}
     * Thrown if the auth cookie is missing or invalid.
     *
     * @throws {@link InternalError}
     * Thrown if something goes wrong on server side.
     */
    public async listBySatellite(satelliteId: string): Promise<Node[]> {
        const path = `${this.ROOT_PATH}/infos/${satelliteId}`;
        const response = await this.http.get(path);

        if (!response.ok) {
            await this.handleError(response);
        }

        const nodeListJson = await response.json();

        return nodeListJson.map(node => new Node(
            node.id,
            node.name,
            node.version,
            new Date(node.lastContact),
            0,
            0,
            0,
            node.onlineScore,
            node.auditScore,
            node.suspensionScore,
            node.totalEarned,
        ));
    }

    /**
     * updates nodes name.
     *
     * @param id - id of the node.
     * @param name - new node name.
     *
     * @throws {@link BadRequestError}
     * This exception is thrown if the input is not a valid.
     *
     * @throws {@link UnauthorizedError}
     * Thrown if the auth cookie is missing or invalid.
     *
     * @throws {@link InternalError}
     * Thrown if something goes wrong on server side.
     */
    public async updateName(id: string, name: string): Promise<void> {
        const path = `${this.ROOT_PATH}/${id}`;
        const response = await this.http.patch(path, JSON.stringify({name: name}));

        if (!response.ok) {
            await this.handleError(response);
        }
    }

    /**
     * deletes node.
     *
     * @param id - id of the node.
     *
     * @throws {@link BadRequestError}
     * This exception is thrown if the input is not a valid.
     *
     * @throws {@link UnauthorizedError}
     * Thrown if the auth cookie is missing or invalid.
     *
     * @throws {@link InternalError}
     * Thrown if something goes wrong on server side.
     */
    public async delete(id: string): Promise<void> {
        const path = `${this.ROOT_PATH}/${id}`;
        const response = await this.http.delete(path);

        if (!response.ok) {
            await this.handleError(response);
        }
    }

    /**
     * retrieves list of trusted satellites node urls for a node.
     */
    public async trustedSatellites(): Promise<NodeURL[]> {
        const path = `${this.ROOT_PATH}/trusted-satellites`;
        const response = await this.http.get(path);

        if (!response.ok) {
            await this.handleError(response);
        }

        const urlListJson = await response.json();

        return urlListJson.map(url => new NodeURL(
            url.ID,
            url.Name,
        ));
    }

    /**
     * handles error due to response code.
     * @param response - response from server.
     *
     * @throws {@link BadRequestError}
     * This exception is thrown if the input is not a valid ISBN number.
     *
     * @throws {@link UnauthorizedError}
     * Thrown if the ISBN number is valid, but no such book exists in the catalog.
     *
     * @throws {@link InternalError}
     * Thrown if the ISBN number is valid, but no such book exists in the catalog.
     *
     * @private
     */
    private async handleError(response: Response): Promise<void> {
        const body = await response.json();

        switch (response.status) {
            case 401: throw new UnauthorizedError(body.error);
            case 400: throw new BadRequestError(body.error);
            case 500:
            default:
                throw new InternalError(body.error);
        }
    }
}
