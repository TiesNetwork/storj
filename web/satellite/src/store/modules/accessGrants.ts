// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import { StoreModule } from '@/store';
import {
    AccessGrant,
    AccessGrantCursor,
    AccessGrantsApi,
    AccessGrantsOrderBy,
    AccessGrantsPage,
    DurationPermission,
    GatewayCredentials,
} from '@/types/accessGrants';
import { SortDirection } from '@/types/common';

export const ACCESS_GRANTS_ACTIONS = {
    FETCH: 'fetchAccessGrants',
    CREATE: 'createAccessGrant',
    DELETE: 'deleteAccessGrants',
    CLEAR: 'clearAccessGrants',
    GET_GATEWAY_CREDENTIALS: 'getGatewayCredentials',
    SET_ACCESS_GRANTS_WEB_WORKER: 'setAccessGrantsWebWorker',
    STOP_ACCESS_GRANTS_WEB_WORKER: 'stopAccessGrantsWebWorker',
    SET_SEARCH_QUERY: 'setAccessGrantsSearchQuery',
    SET_SORT_BY: 'setAccessGrantsSortingBy',
    SET_SORT_DIRECTION: 'setAccessGrantsSortingDirection',
    SET_DURATION_PERMISSION: 'setAccessGrantsDurationPermission',
    TOGGLE_SELECTION: 'toggleAccessGrantsSelection',
    TOGGLE_BUCKET_SELECTION: 'toggleBucketSelection',
    CLEAR_SELECTION: 'clearAccessGrantsSelection',
};

export const ACCESS_GRANTS_MUTATIONS = {
    SET_PAGE: 'setAccessGrants',
    SET_GATEWAY_CREDENTIALS: 'setGatewayCredentials',
    SET_ACCESS_GRANTS_WEB_WORKER: 'setAccessGrantsWebWorker',
    STOP_ACCESS_GRANTS_WEB_WORKER: 'stopAccessGrantsWebWorker',
    TOGGLE_SELECTION: 'toggleAccessGrantsSelection',
    TOGGLE_BUCKET_SELECTION: 'toggleBucketSelection',
    CLEAR_SELECTION: 'clearAccessGrantsSelection',
    CLEAR: 'clearAccessGrants',
    CHANGE_SORT_ORDER: 'changeAccessGrantsSortOrder',
    CHANGE_SORT_ORDER_DIRECTION: 'changeAccessGrantsSortOrderDirection',
    SET_SEARCH_QUERY: 'setAccessGrantsSearchQuery',
    SET_PAGE_NUMBER: 'setAccessGrantsPage',
    SET_DURATION_PERMISSION: 'setAccessGrantsDurationPermission',
};

const {
    SET_PAGE,
    TOGGLE_SELECTION,
    TOGGLE_BUCKET_SELECTION,
    CLEAR_SELECTION,
    CLEAR,
    CHANGE_SORT_ORDER,
    CHANGE_SORT_ORDER_DIRECTION,
    SET_SEARCH_QUERY,
    SET_PAGE_NUMBER,
    SET_DURATION_PERMISSION,
    SET_GATEWAY_CREDENTIALS,
    SET_ACCESS_GRANTS_WEB_WORKER,
    STOP_ACCESS_GRANTS_WEB_WORKER,
} = ACCESS_GRANTS_MUTATIONS;

export class AccessGrantsState {
    public cursor: AccessGrantCursor = new AccessGrantCursor();
    public page: AccessGrantsPage = new AccessGrantsPage();
    public selectedAccessGrantsIds: string[] = [];
    public selectedBucketNames: string[] = [];
    public permissionNotBefore: Date = new Date();
    public permissionNotAfter: Date = new Date('2200-01-01');
    public gatewayCredentials: GatewayCredentials = new GatewayCredentials();
    public accessGrantsWebWorker: Worker | null = null;
    public isAccessGrantsWebWorkerReady: boolean = false;
}

/**
 * creates access grants module with all dependencies
 *
 * @param api - accessGrants api
 */
export function makeAccessGrantsModule(api: AccessGrantsApi): StoreModule<AccessGrantsState> {
    return {
        state: new AccessGrantsState(),
        mutations: {
            [SET_ACCESS_GRANTS_WEB_WORKER](state: AccessGrantsState): void {
                state.accessGrantsWebWorker = new Worker('@/../static/wasm/accessGrant.worker.js', { type: 'module' });
                state.accessGrantsWebWorker.onmessage = (event: MessageEvent) => {
                    const data = event.data;
                    if (data !== 'configured') {
                        console.error('Failed to configure access grants web worker');

                        return;
                    }

                    state.isAccessGrantsWebWorkerReady = true;
                };
                state.accessGrantsWebWorker.onerror = (error: ErrorEvent) => {
                    console.error(`Failed to configure access grants web worker. ${error.message}`);
                };
            },
            [STOP_ACCESS_GRANTS_WEB_WORKER](state: AccessGrantsState): void {
                state.accessGrantsWebWorker?.postMessage({
                    'type': 'Stop',
                });
                state.accessGrantsWebWorker = null;
                state.isAccessGrantsWebWorkerReady = false;
            },
            [SET_PAGE](state: AccessGrantsState, page: AccessGrantsPage) {
                state.page = page;
                state.page.accessGrants = state.page.accessGrants.map(accessGrant => {
                    if (state.selectedAccessGrantsIds.includes(accessGrant.id)) {
                        accessGrant.isSelected = true;
                    }

                    return accessGrant;
                });
            },
            [SET_GATEWAY_CREDENTIALS](state: AccessGrantsState, credentials: GatewayCredentials) {
                state.gatewayCredentials = credentials;
            },
            [SET_PAGE_NUMBER](state: AccessGrantsState, pageNumber: number) {
                state.cursor.page = pageNumber;
            },
            [SET_SEARCH_QUERY](state: AccessGrantsState, search: string) {
                state.cursor.search = search;
            },
            [SET_DURATION_PERMISSION](state: AccessGrantsState, permission: DurationPermission) {
                state.permissionNotBefore = permission.notBefore;
                state.permissionNotAfter = permission.notAfter;
            },
            [CHANGE_SORT_ORDER](state: AccessGrantsState, order: AccessGrantsOrderBy) {
                state.cursor.order = order;
            },
            [CHANGE_SORT_ORDER_DIRECTION](state: AccessGrantsState, direction: SortDirection) {
                state.cursor.orderDirection = direction;
            },
            [TOGGLE_SELECTION](state: AccessGrantsState, accessGrant: AccessGrant) {
                if (!state.selectedAccessGrantsIds.includes(accessGrant.id)) {
                    state.page.accessGrants.forEach((grant: AccessGrant) => {
                        if (grant.id === accessGrant.id) {
                            grant.isSelected = true;
                        }
                    });
                    state.selectedAccessGrantsIds.push(accessGrant.id);

                    return;
                }

                state.page.accessGrants.forEach((grant: AccessGrant) => {
                    if (grant.id === accessGrant.id) {
                        grant.isSelected = false;
                    }
                });
                state.selectedAccessGrantsIds = state.selectedAccessGrantsIds.filter(accessGrantId => {
                    return accessGrant.id !== accessGrantId;
                });
            },
            [TOGGLE_BUCKET_SELECTION](state: AccessGrantsState, bucketName: string) {
                if (!state.selectedBucketNames.includes(bucketName)) {
                    state.selectedBucketNames.push(bucketName);

                    return;
                }

                state.selectedBucketNames = state.selectedBucketNames.filter(name => {
                    return bucketName !== name;
                });
            },
            [CLEAR_SELECTION](state: AccessGrantsState) {
                state.selectedBucketNames = [];
                state.selectedAccessGrantsIds = [];
                state.page.accessGrants = state.page.accessGrants.map((accessGrant: AccessGrant) => {
                    accessGrant.isSelected = false;

                    return accessGrant;
                });
            },
            [CLEAR](state: AccessGrantsState) {
                state.cursor = new AccessGrantCursor();
                state.page = new AccessGrantsPage();
                state.selectedAccessGrantsIds = [];
                state.selectedBucketNames = [];
                state.permissionNotBefore = new Date();
                state.permissionNotAfter = new Date('2200-01-01');
                state.gatewayCredentials = new GatewayCredentials();
            },
        },
        actions: {
            setAccessGrantsWebWorker: function({commit}: any): void {
                commit(SET_ACCESS_GRANTS_WEB_WORKER);
            },
            stopAccessGrantsWebWorker: function({commit}: any): void {
                commit(STOP_ACCESS_GRANTS_WEB_WORKER);
            },
            fetchAccessGrants: async function ({commit, rootGetters, state}, pageNumber: number): Promise<AccessGrantsPage> {
                const projectId = rootGetters.selectedProject.id;
                commit(SET_PAGE_NUMBER, pageNumber);

                const accessGrantsPage: AccessGrantsPage = await api.get(projectId, state.cursor);
                commit(SET_PAGE, accessGrantsPage);

                return accessGrantsPage;
            },
            createAccessGrant: async function ({commit, rootGetters}: any, name: string): Promise<AccessGrant> {
                const accessGrant = await api.create(rootGetters.selectedProject.id, name);

                return accessGrant;
            },
            deleteAccessGrants: async function({state}: any): Promise<void> {
                await api.delete(state.selectedAccessGrantsIds);
            },
            getGatewayCredentials: async function({state, commit}: any, accessGrant: string): Promise<void> {
                const credentials: GatewayCredentials = await api.getGatewayCredentials(accessGrant);

                commit(SET_GATEWAY_CREDENTIALS, credentials);
            },
            setAccessGrantsSearchQuery: function ({commit}, search: string) {
                commit(SET_SEARCH_QUERY, search);
            },
            setAccessGrantsSortingBy: function ({commit}, order: AccessGrantsOrderBy) {
                commit(CHANGE_SORT_ORDER, order);
            },
            setAccessGrantsSortingDirection: function ({commit}, direction: SortDirection) {
                commit(CHANGE_SORT_ORDER_DIRECTION, direction);
            },
            setAccessGrantsDurationPermission: function ({commit}, permission: DurationPermission) {
                commit(SET_DURATION_PERMISSION, permission);
            },
            toggleAccessGrantsSelection: function ({commit}, accessGrant: AccessGrant): void {
                commit(TOGGLE_SELECTION, accessGrant);
            },
            toggleBucketSelection: function ({commit}, bucketName: string): void {
                commit(TOGGLE_BUCKET_SELECTION, bucketName);
            },
            clearAccessGrantsSelection: function ({commit}): void {
                commit(CLEAR_SELECTION);
            },
            clearAccessGrants: function ({commit}): void {
                commit(CLEAR);
                commit(CLEAR_SELECTION);
            },
        },
        getters: {
            selectedAccessGrants: (state: AccessGrantsState) => state.page.accessGrants.filter((grant: AccessGrant) => grant.isSelected),
        },
    };
}
