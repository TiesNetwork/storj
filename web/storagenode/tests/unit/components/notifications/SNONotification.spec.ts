// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import Vuex from 'vuex';

import SNONotification from '@/app/components/notifications/SNONotification.vue';

import { newNotificationsModule } from '@/app/store/modules/notifications';
import { UINotification } from '@/app/types/notifications';
import { NotificationsHttpApi } from '@/storagenode/api/notifications';
import { Notification, NotificationTypes } from '@/storagenode/notifications/notifications';
import { NotificationsService } from '@/storagenode/notifications/service';
import { createLocalVue, shallowMount } from '@vue/test-utils';

const localVue = createLocalVue();
localVue.use(Vuex);

const notificationsApi = new NotificationsHttpApi();
const notificationsService = new NotificationsService(notificationsApi);
const notificationsModule = newNotificationsModule(notificationsService);

const store = new Vuex.Store({ modules: { notificationsModule }});

describe('SNONotification', (): void => {
    it('renders correctly with default props', (): void => {
        const wrapper = shallowMount(SNONotification, {
            store,
            localVue,
        });

        expect(wrapper).toMatchSnapshot();
    });

    it('renders correctly', async (): Promise<void> => {
        const mockMethod = jest.fn();
        const testNotification = new Notification(
            '123',
            '1234',
            NotificationTypes.AuditCheckFailure,
            'title1',
            'message1',
        );

        const wrapper = shallowMount(SNONotification, {
            store,
            localVue,
            propsData: {
                notification: new UINotification(testNotification),
                isSmall: false,
            },
        });

        wrapper.setMethods({ read: mockMethod });

        await wrapper.find('.notification-item').trigger('mouseenter');

        expect(mockMethod).toHaveBeenCalled();
        expect(wrapper).toMatchSnapshot();
    });
});
