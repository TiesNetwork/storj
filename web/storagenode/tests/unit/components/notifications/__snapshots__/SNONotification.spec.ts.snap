// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SNONotification renders correctly 1`] = `
<div class="notification-item unread">
  <div class="row">
    <div class="notification-item__new-indicator-container"><span class="notification-item__new-indicator-container__circle"></span></div>
    <div class="notification-item__icon-container">
      <div class="icon"><svg width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
          <path class="close-cross-svg-path" d="M15.7071 1.70711C16.0976 1.31658 16.0976 0.683417 15.7071 0.292893C15.3166 -0.0976311 14.6834 -0.0976311 14.2929 0.292893L15.7071 1.70711ZM0.292893 14.2929C-0.0976311 14.6834 -0.0976311 15.3166 0.292893 15.7071C0.683417 16.0976 1.31658 16.0976 1.70711 15.7071L0.292893 14.2929ZM1.70711 0.292893C1.31658 -0.0976311 0.683417 -0.0976311 0.292893 0.292893C-0.0976311 0.683417 -0.0976311 1.31658 0.292893 1.70711L1.70711 0.292893ZM14.2929 15.7071C14.6834 16.0976 15.3166 16.0976 15.7071 15.7071C16.0976 15.3166 16.0976 14.6834 15.7071 14.2929L14.2929 15.7071ZM14.2929 0.292893L0.292893 14.2929L1.70711 15.7071L15.7071 1.70711L14.2929 0.292893ZM0.292893 1.70711L14.2929 15.7071L15.7071 14.2929L1.70711 0.292893L0.292893 1.70711Z" fill="#384B65"></path>
        </svg>
      </div>
    </div>
    <div class="notification-item__text-container">
      <p class="notification-item__text-container__message"><b class="notification-item__text-container__message__bold">title1:</b> message1
      </p>
      <!---->
    </div>
  </div>
  <div class="notification-item__date-container">
    <p class="notification-item__date-container__date">Just now</p>
  </div>
</div>
`;

exports[`SNONotification renders correctly with default props 1`] = `
<div class="notification-item unread">
  <div class="row">
    <div class="notification-item__new-indicator-container"><span class="notification-item__new-indicator-container__circle"></span></div>
    <div class="notification-item__icon-container">
      <div class="icon"><svg width="15" height="18" viewBox="0 0 15 18" fill="none" xmlns="http://www.w3.org/2000/svg">
          <path fill-rule="evenodd" clip-rule="evenodd" d="M6.09394 2.14899C3.66843 2.80225 1.87523 5.10849 1.87523 7.85462V11.7819C1.87523 12.1437 1.59447 12.4364 1.25145 12.4364C0.560552 12.4364 0.000244141 13.0245 0.000244141 13.7455C0.000244141 14.469 0.560552 15.0546 1.25389 15.0546H13.7466C14.4387 15.0546 15.0002 14.4665 15.0002 13.7455C15.0002 13.0219 14.4375 12.4364 13.749 12.4364C13.4048 12.4364 13.1253 12.1449 13.1253 11.7819V7.85462C13.1253 5.1099 11.3333 2.80353 8.90655 2.14899V1.47272C8.90655 0.657093 8.27668 0 7.50031 0C6.72516 0 6.09407 0.659646 6.09407 1.47272L6.09394 2.14899ZM5.3127 15.7091H9.68766C9.68766 16.9747 8.70867 18 7.50018 18C6.29169 18 5.3127 16.9747 5.3127 15.7091Z" fill="#535F77"></path>
        </svg></div>
    </div>
    <div class="notification-item__text-container">
      <p class="notification-item__text-container__message"><b class="notification-item__text-container__message__bold">:</b>
      </p>
      <!---->
    </div>
  </div>
  <div class="notification-item__date-container">
    <p class="notification-item__date-container__date">Just now</p>
  </div>
</div>
`;
